src/Microsoft.AspNetCore.Server.IISIntegration/Server/IISHttpContext.cs

```diff
@@ -258,24 +239,266 @@ private IISAwaitable DoFlushAsync()
+                    if (_readWriteTask == null)
+                    {
+                        _readWriteTask = ConsumeAsync();
+                    }
+                }
+            }
+        }
+
+        // ConsumeAsync is called when either the first read or first write is done. 
+        // There are two modes for reading and writing to the request/response bodies.
+        // 1. Await all reads and try to read from the Output pipe
+        // 2. Done reading and await all writes.
+        public async Task ConsumeAsync()
+        {
+            await ReadAndWriteLoopAsync();
+
+            // If we are done writing, complete the output pipe and return
+            // Input Pipe will be closed when ReadAndWriteLoopAsync returns
+            if (_doneWriting)
+            {
+                Output.Reader.Complete();
+                return;
+            }
+
+            await WriteLoopAsync();
+        }
+
+        public async Task ReadAndWriteLoopAsync()
+        {
+            // First we check if there is anything to write from the Output pipe
+            // If there is, we call WriteToIISAsync
+            while (!_doneReading)
+            {
+                // Check if Output pipe has anything to write to IIS.
+                if (Output.Reader.TryRead(out var readResult))
+                {
+                    var buffer = readResult.Buffer;
+                    var consumed = buffer.End;
+
+                    try
+                    {
+                        // If the output pipe is termniated, cancel the request.
+                        if (readResult.IsCancelled)
+                        {
+                            _doneWriting = true;
+                            break;
+                        }
+
+                        if (!buffer.IsEmpty)
+                        {
+                            // Write to IIS buffers
+                            // Guaranteed to write the entire buffer to IIS
+                            await WriteToIISAsync(buffer);
+                        }
+                        else if (readResult.IsCompleted)
+                        {
+                            break;
+                        }
+                        else
+                        {
+                            // Flush of zero bytes
+                            await FlushToIISAsync();
+                        }
+                    }
+                    catch (Exception ex)
+                    {
+                        throw ex;
+                    }
+                    finally
+                    {
+                        // Always Advance the data pointer to the end of the buffer.
+                        Output.Reader.AdvanceTo(buffer.End);
+                    }
+                }
+
+               
+                // Check if there was an upgrade. If there is, we will replace the request and response bodies with
+                // two seperate loops
+                await CheckForUpgrade();
```

Comment:
Won't this loop Complete the input and output writer and reader? Won't that make the code below fail as it tries to access memory after completing?

Reply:
You are correct. There is also the inverse issue in the WriteLoop where the Input will already be completed.